### PR TITLE
feat(admin): Add DEX-describing data fields

### PR DESCRIPTION
**Pull request by KirilMihaylov**

### CODE REVIEW

File: platform/contracts/admin/src/contracts/protocol/mod.rs [hunk 1]
```diff
@@ -22,5 +22,7 @@ pub struct ProtocolTemplate<T> {
 #[serde(rename_all = "snake_case", deny_unknown_fields)]
 pub struct Protocol {
     pub network: String,
+    pub dex_type: String,
+    pub dex_metadata: String,
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/protocol/mod.rs:26]: we've discussed alternative naming, please use it

File: platform/contracts/admin/src/contracts/protocol/mod.rs [hunk 2]
```diff
@@ -22,5 +22,7 @@ pub struct ProtocolTemplate<T> {
 #[serde(rename_all = "snake_case", deny_unknown_fields)]
 pub struct Protocol {
     pub network: String,
+    pub dex_type: String,
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/protocol/mod.rs:25]: using `String`s here is not adequate .... we have to restrict to the supported ones

File: platform/contracts/admin/src/endpoints.rs [hunk 3]
```diff
@@ -49,13 +48,12 @@ pub fn instantiate(
...
 }
 
-#[cfg(feature = "migrate")]
 #[entry_point]
 pub fn migrate(
     deps: DepsMut<'_>,
     env: Env,
     MigrateMsg {
-        reserve_contracts,
+        dex_data,
```
Comment (gmanev7) [inline platform/contracts/admin/src/endpoints.rs:56]: do you believe that all existing protocols share the same dex_data? Or I am mistaken here?

Comment (gmanev7) [reply platform/contracts/admin/src/endpoints.rs:56]: disregard it, just saw the type

File: platform/contracts/admin/src/msg.rs [hunk 4]
```diff
@@ -89,6 +87,13 @@ pub enum QueryMsg {
     Protocol(String),
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct DexData {
+    pub dex_type: String,
+    pub dex_params: Option<String>,
```
Comment (gmanev7) [inline platform/contracts/admin/src/msg.rs:94]: please remove the prefix `dex_`

File: platform/contracts/admin/src/contracts/protocol/mod.rs [hunk 5]
```diff
@@ -21,6 +21,57 @@ pub struct ProtocolTemplate<T> {
...
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[serde(
+    rename_all = "PascalCase",
+    rename_all_fields = "snake_case",
+    deny_unknown_fields
+)]
+pub enum Network {
+    #[serde(alias = "NEUTRON")]
+    Neutron,
+    #[serde(alias = "OSMOSIS")]
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/protocol/mod.rs:38]: what's the idea of having aliases? They do allow supporting a second name, but only on deserialization.

File: platform/contracts/admin/src/contracts/protocol/mod.rs [hunk 6]
```diff
@@ -21,6 +21,57 @@ pub struct ProtocolTemplate<T> {
...
-    pub network: String,
+    pub network: Network,
+    pub dex: Dex,
     pub contracts: ProtocolTemplate<Addr>,
 }
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[serde(
+    rename_all = "PascalCase",
+    rename_all_fields = "snake_case",
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/protocol/mod.rs:32]: as per the [doc](https://serde.rs/container-attrs.html) these two do the same when applied on enums, as it is in our case ..... and you have specified two different styles which would confuse the reader

File: platform/contracts/admin/src/contracts/protocol/mod.rs [hunk 7]
```diff
@@ -21,6 +21,57 @@ pub struct ProtocolTemplate<T> {
...
+    #[serde(alias = "NEUTRON")]
+    Neutron,
+    #[serde(alias = "OSMOSIS")]
+    Osmosis,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[serde(
+    rename_all = "PascalCase",
+    rename_all_fields = "snake_case",
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/protocol/mod.rs:45]: same here

File: platform/contracts/admin/src/contracts/mod.rs [hunk 8]
```diff
@@ -32,7 +32,7 @@ pub type ContractsMigration = ContractsTemplate<Option<MigrationSpec>>;
 
 pub type ContractsPostMigrationExecute = ContractsTemplate<Option<String>>;
 
-pub(crate) type ContractsGroupedByProtocol = ContractsTemplate<Addr, BTreeMap<String, Protocol>>;
+pub type ContractsGroupedByProtocol = ContractsTemplate<Addr, BTreeMap<String, Protocol>>;
```
Comment (gmanev7) [inline platform/contracts/admin/src/contracts/mod.rs:35]: who is gonna use it outside the crate?

File: platform/contracts/admin/src/state/contracts.rs [hunk 9]
```diff
@@ -76,12 +74,9 @@ pub(crate) fn load_all(storage: &dyn Storage) -> Result<ContractsGroupedByProtoc
     })
 }
 
-#[cfg(feature = "migrate")]
 pub(super) fn migrate_protocols(
     storage: &mut dyn Storage,
-    api: &dyn Api,
-    querier: QuerierWrapper<'_>,
-    mut reserve_contracts: BTreeMap<String, String>,
+    mut dex_types: BTreeMap<String, Dex>,
```
Comment (gmanev7) [inline platform/contracts/admin/src/state/contracts.rs:79]: [globally] please rename `dex_types` to `dexes`

File: platform/contracts/admin/src/state/contracts.rs [hunk 10]
```diff
@@ -91,56 +86,30 @@ pub(super) fn migrate_protocols(
...
-    pub network: String,
-    pub contracts: OldProtocolTemplate<Addr>,
+    pub network: Network,
+    pub contracts: ProtocolTemplate<Addr>,
 }
 
-#[cfg(feature = "migrate")]
 impl OldProtocol {
-    fn convert(self, reserve: Addr) -> Protocol {
+    fn convert(self, dex: Dex) -> Protocol {
```
Comment (gmanev7) [inline platform/contracts/admin/src/state/contracts.rs:108]: `convert` -> `migrate`

File: platform/contracts/admin/src/contracts/protocol/mod.rs (hunk 5 above)
Comment (KirilMihaylov) [reply platform/contracts/admin/src/contracts/protocol/mod.rs:38]: I put them specifically for this migration. Later on we'll remove them, but right now we need them in order to automatically convert them from `SCREAMING_SNAKE_CASE` to `PascalCase`.

Comment (KirilMihaylov) [reply platform/contracts/admin/src/contracts/protocol/mod.rs:38]: I deemed that testing well-known `serde` attributes is beyond our scope. I had put tests on the mixed casing as we haven't used that before and it is after all mixing them, that behaviors is not explicitly documented.

Comment (gmanev7) [reply platform/contracts/admin/src/contracts/protocol/mod.rs:38]: OK, then put a TODO around to not forget their purpose and when supposed to be removed

File: platform/contracts/admin/src/contracts/mod.rs (hunk 8 above)
Comment (KirilMihaylov) [reply platform/contracts/admin/src/contracts/mod.rs:35]: It is used outside, in the `InstantiateMsg`, if I recall correctly. The compiler was livid about it.

File: platform/contracts/admin/src/contracts/protocol/mod.rs (hunk 6 above)
Comment (KirilMihaylov) [reply platform/contracts/admin/src/contracts/protocol/mod.rs:32]: I have a test covering it. It absolutely does NOT do the same thing when they are combined. I've applied them both in order to ensure consistency of the casing of the fields themselves.

Comment (gmanev7) [reply platform/contracts/admin/src/contracts/protocol/mod.rs:32]: my bad reading through the doc

**Review by gmanev7 — APPROVED** [review]

**Pull request merged**